EIPS/eip-5988.md

```diff
@@ -0,0 +1,163 @@
+
+ZK-Rollups have particular needs for cryptographic hash functions that can allow for efficient verification of proofs.
+
+The Poseidon hash function is a set of permutations over a prime field, which makes it particularly well-suited for the purpose of building efficient ZK / Validity rollups on Ethereum.
+
+Poseidon is one of the most efficient hashing algorithms that can be used in this context.
+Moreover it is compatible with all major proof systems (SNARKs, STARKs, Bulletproofs, etc...).
+This makes it a good candidate for a precompile that can be used by many different ZK-Rollups.
+
+An important point to note is that ZK rollups using Poseidon have chosen different sets of parameters, which makes it harder to build a single precompile for all of them.
+
+However, we can still build a generic precompile that supports arbitrary parameters, and allow the ZK rollups to choose the parameters they want to use.
+
+This is the approach that we have taken in this EIP.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant                      | Value |
+| ----------------------------- | ----- |
+| `FORK_BLKNUM`                 | `TBD` |
+| `GAS_COST`                    | `TBD` |
+| `POSEIDON_PRECOMPILE_ADDRESS` | `0xA` |
+
+Here are the Poseidon parameters that the precompile will support:
+
+- `p`: the prime field modulus.
+- `security_level`: the security level measured in bits. Denoted `M` in the Poseidon paper.
+- `alpha`: the power of S-box.
+- `input_rate`: the size of input.
+- `t`: the size of the state.
+- `full_round`: the number of full rounds. Denoted as `R_F` in the Poseidon paper.
+- `partial_round`: the number of partial rounds. Denoted as `R_P` in the Poseidon paper.
+
+The input to the precompile MUST be encoded as follows:
+
+- `p`: 32 bytes
+- `security_level`: 2 bytes
+- `alpha`: 1 byte
+- `input_rate`: 2 bytes
+- `t`: 1 byte
+- `full_round`: 1 byte
+- `partial_round`: 1 byte
+- `input`: `input_rate` \* 32 bytes
+
+The encoding of the precompile input is the following:
+
+```text
+[32 bytes for p][2 bytes for security_level][1 byte for alpha][2 bytes for input_rate][1 byte for t][1 byte for full_round][1 byte for partial_round][input_rate * 32 bytes for input]
+```
+
+The precompile should compute the hash function as [specified in the Poseidon paper](../assets/eip-5988/papers/poseidon_paper.pdf) and return hash output.
+
+<!--### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `poseidon_hash` function.
+
+```solidity
+// TODO: Add solidity example
+```-->
+
+<!--### Gas Costs
+
+```text
+TODO: Fill gas costs section and add benchmarks
```

Comment:
```suggestion
TODO: Fill gas costs section
```
Gas cost shall exist in Specification section. I believe benchmarks and other computation cost estimations shall live in Rationale section, considering splitting this TODO here and Rationale